File: rust/src/lib/warmup/valid_anagram.rs
use std::collections::HashMap;

fn is_anagram(s: &str, t: &str) -> bool {
    if s.len() != t.len() {
        return false;
    }

    let mut counter: HashMap<char, i32> = HashMap::new();

    s.chars().zip(t.chars()).for_each(|(sc, tc)| {
        *counter.entry(sc).or_insert(0) += 1;
        *counter.entry(tc).or_insert(0) -= 1;
    });

    counter.values().all(|&val| val == 0)
}

#[cfg(test)]
pub mod test {
    use super::*;

    #[test]
    fn test_is_anagram() {
        let test_cases = [
            ("listen", "silent", true),
            ("hello", "world", false),
            ("anagram", "nagaram", true),
            ("rat", "car", false),
            ("", "", true),
        ];

        for (s, t, expected) in test_cases {
            assert_eq!(is_anagram(s, t), expected);
        }
    }
}

File: javascript/src/warmup/validAnagram.js
export function isAnagram(s, t) {
    if (s.length !== t.length) {
        return false;
    }

    const counter = {};

    for (let i = 0; i < s.length; i++) {
        counter[s[i]] = (counter[s[i]] || 0) + 1;
        counter[t[i]] = (counter[t[i]] || 0) - 1;
    }

    for (const val of Object.values(counter)) {
        if (val != 0) {
            return false;
        }
    }

    return true;
}
